File: source/FloodFill.cpp
// https://leetcode.com/problems/flood-fill
#include <vector>

class Solution
{
public:
    void ColorNeighbor(std::vector<std::vector<int>>& image, int sr, int sc, int newColor, int originalColor)
    {
        if (sr < 0 || sr >= image.size() || sc < 0 || sc >= image[0].size()) return;
        if (image[sr][sc] == newColor) return;

        if (image[sr][sc] == originalColor)
        {
            image[sr][sc] = newColor;

            ColorNeighbor(image, sr - 1, sc, newColor, originalColor);
            ColorNeighbor(image, sr + 1, sc, newColor, originalColor);
            ColorNeighbor(image, sr, sc - 1, newColor, originalColor);
            ColorNeighbor(image, sr, sc + 1, newColor, originalColor);
        }
    }

    std::vector<std::vector<int>> floodFill(std::vector<std::vector<int>>& image, int sr, int sc, int color)
    {
        if (image.empty()) return image;

        ColorNeighbor(image, sr - 1, sc, color, image[sr][sc]);
        ColorNeighbor(image, sr + 1, sc, color, image[sr][sc]);
        ColorNeighbor(image, sr, sc - 1, color, image[sr][sc]);
        ColorNeighbor(image, sr, sc + 1, color, image[sr][sc]);

        image[sr][sc] = color;

        return image;
    }
};

File: CSharpPractice/source/FloodFill.cs
// https://leetcode.com/problems/flood-fill
using System.Collections.Generic;

namespace LeetCodePractice.FloodFill
{
    public class Solution
    {
        public int[][] FloodFill(int[][] image, int sr, int sc, int color)
        {
            int originalColor = image[sr][sc];
            if (originalColor == color) return image;

            int rows = image.Length;
            int cols = image[0].Length;

            Queue<(int, int)> queue = new Queue<(int, int)>();
            queue.Enqueue((sr, sc));

            while (queue.Count > 0)
            {
                var (row, col) = queue.Dequeue();
                image[row][col] = color;

                if (row - 1 >= 0 && image[row - 1][col] == originalColor) queue.Enqueue((row - 1, col));
                if (row + 1 < rows && image[row + 1][col] == originalColor) queue.Enqueue((row + 1, col));
                if (col - 1 >= 0 && image[row][col - 1] == originalColor) queue.Enqueue((row, col - 1));
                if (col + 1 < cols && image[row][col + 1] == originalColor) queue.Enqueue((row, col + 1));
            }

            return image;
        }
    }
}